jetty-http/src/main/java/org/eclipse/jetty/http/HttpMethod.java

```diff
@@ -26,162 +26,192 @@
 
-    private final ByteBuffer _buffer;
-    private final byte[] _bytes;
+    public ByteBuffer asBuffer()
+    {
+        return _buffer.asReadOnlyBuffer();
+    }
 
-    HttpMethod()
+    public String asString()
     {
-        _bytes = StringUtil.getBytes(toString());
-        _buffer = ByteBuffer.wrap(_bytes);
+        return _method;
     }
 
-    public byte[] getBytes()
+    public String toString()
     {
-        return _bytes;
+        return _method;
     }
 
-    public boolean is(String s)
+    public static final Trie<HttpMethod> INSENSITIVE_CACHE = new ArrayTrie<>(252);
+    public static final Trie<HttpMethod> CACHE = new ArrayTernaryTrie<>(false, 300);
+    public static final Trie<HttpMethod> LOOK_AHEAD = new ArrayTernaryTrie<>(false, 330);
+    private static final int ACL_AS_INT = ('A' & 0xff) << 24 | ('C' & 0xFF) << 16 | ('L' & 0xFF) << 8 | (' ' & 0xFF);
+    private static final int GET_AS_INT = ('G' & 0xff) << 24 | ('E' & 0xFF) << 16 | ('T' & 0xFF) << 8 | (' ' & 0xFF);
+    private static final int PRI_AS_INT = ('P' & 0xff) << 24 | ('R' & 0xFF) << 16 | ('I' & 0xFF) << 8 | (' ' & 0xFF);
+    private static final int PUT_AS_INT = ('P' & 0xff) << 24 | ('U' & 0xFF) << 16 | ('T' & 0xFF) << 8 | (' ' & 0xFF);
+    static
     {
-        return toString().equalsIgnoreCase(s);
+        for (HttpMethod method : HttpMethod.values())
+        {
+            if (!INSENSITIVE_CACHE.put(method.asString(), method))
+                throw new IllegalStateException("INSENSITIVE_CACHE too small: " + method);
+
+            if (!CACHE.put(method.asString(), method))
+                throw new IllegalStateException("CACHE too small: " + method);
+
+            if (!LOOK_AHEAD.put(method.asString() + ' ', method))
+                throw new IllegalStateException("LOOK_AHEAD too small: " + method);
+        }
     }
 
-    public ByteBuffer asBuffer()
+    /**
+     * Optimized lookup to find a method name and trailing space in a byte array.
+     *
+     * @param bytes Array containing ISO-8859-1 characters
+     * @param position The first valid index
+     * @param limit The first non valid index
+     * @return An HttpMethod if a match or null if no easy match.
+     * @deprecated Not used
+     */
+    @Deprecated
+    public static HttpMethod lookAheadGet(byte[] bytes, final int position, int limit)
     {
-        return _buffer.asReadOnlyBuffer();
+        return LOOK_AHEAD.getBest(bytes, position, limit - position);
     }
 
-    public String asString()
+    /**
+     * Optimized lookup to find a method name and trailing space in a byte array.
+     *
+     * @param buffer buffer containing ISO-8859-1 characters, it is not modified.
+     * @return An HttpMethod if a match or null if no easy match.
+     */
+    public static HttpMethod lookAheadGet(ByteBuffer buffer)
     {
-        return toString();
+        int len = buffer.remaining();
+        // Short cut for 3 char methods, mostly for GET optimisation
+        if (len > 3)
```

Comment:
Is this really necessary?
Does this really produce any measurable performance improvement? (have you jmh'd this?)

Reply:
The getInt switch is an order of magnitude better.  Interestingly the Map is slightly better than the Trie in this case sensitive case, but the map needs to allocate.
So doing this benchmark has convinced me I should do the getInt trick for POST and HEAD as well!

```
Benchmark                                                     Mode  Cnt          Score         Error   Units
HttpMethodBenchmark.testIntSwitch                            thrpt   10  234190385.517 ± 1954845.858   ops/s
HttpMethodBenchmark.testIntSwitch:·gc.alloc.rate.norm        thrpt   10         ≈ 10⁻⁵                  B/op
HttpMethodBenchmark.testMapGet                               thrpt   10   21957916.326 ±  507531.533   ops/s
HttpMethodBenchmark.testMapGet:·gc.alloc.rate.norm           thrpt   10         48.000 ±       0.001    B/op
HttpMethodBenchmark.testTrieGetBest                          thrpt   10   18475001.742 ±  205942.699   ops/s
HttpMethodBenchmark.testTrieGetBest:·gc.alloc.rate.norm      thrpt   10         ≈ 10⁻⁴                  B/op

```

Reply:
Extending the int switch to POST and benchmarking against MOVE which doesn't have the optimisation gives:
```
Benchmark                                                     Mode  Cnt          Score          Error   Units
HttpMethodBenchmark.testHttpMethodMove                       thrpt   10   10435118.064 ±   240925.892   ops/s
HttpMethodBenchmark.testHttpMethodMove:·gc.alloc.rate.norm   thrpt   10         ≈ 10⁻⁴                   B/op
HttpMethodBenchmark.testHttpMethodPost                       thrpt   10  148142781.323 ± 14602351.496   ops/s
HttpMethodBenchmark.testHttpMethodPost:·gc.alloc.rate.norm   thrpt   10         ≈ 10⁻⁵                   B/op
HttpMethodBenchmark.testIntSwitch                            thrpt   10  206759099.264 ±  6371501.120   ops/s
HttpMethodBenchmark.testIntSwitch:·gc.alloc.rate.norm        thrpt   10         ≈ 10⁻⁵                   B/op
HttpMethodBenchmark.testMapGet                               thrpt   10   20527043.581 ±   415407.826   ops/s
HttpMethodBenchmark.testMapGet:·gc.alloc.rate.norm           thrpt   10         48.000 ±        0.001    B/op
HttpMethodBenchmark.testTrieGetBest                          thrpt   10   18573224.811 ±   591011.908   ops/s
HttpMethodBenchmark.testTrieGetBest:·gc.alloc.rate.norm      thrpt   10         ≈ 10⁻⁴                   B/op
```

So even the slightly more complex lookup for POST easily beats the non optimised MOVE by an order of magnitude.
Makes me wonder if we shouldn't make Trie (or is it now an Index) that uses getLong and getInt for most of the work!